File: 구름 알고리즘 미션/자료구조/묶음 상품/묶음 상품.py
import sys
input = sys.stdin.readline

def find(u):
	if parent[u] != u:
		parent[u] = find(parent[u])
	return parent[u]

def union(u, v):
	root_u = find(u)
	root_v = find(v)
	
	if root_u == root_v:
		return False
	
	if root_u < root_v:
		parent[root_v] = root_u
	else:
		parent[root_u] = root_v
	return True

N, M = map(int, input().split())
parent = [i for i in range(N+1)]
cnt = N
for _ in range(M):
	a, b = map(int, input().split())
	if union(a, b):
		cnt -= 1
		
print(cnt)

File: 구름 알고리즘 미션/자료구조/묶음 상품/묶음 상품.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

let input = [];
rl.on('line', line => {
  input.push(line.trim());
}).on('close', () => {
	
	const find = (u) => {
		if(parent[u] !== u) parent[u] = find(parent[u]);
		return parent[u];
	}
	
	const union = (u, v) => {
		const rootU = find(u);
		const rootV = find(v);
		
		if(rootU === rootV) return false;
		if(rootU < rootV) parent[rootV] = rootU;
		else parent[rootU] = rootV;
		return true;
	}
	
  const [N, M] = input[0].split(' ').map(Number);
  const parent = Array.from({ length: N + 1 }, (_, i) => i);
	let cnt = N;
	
	for(let i = 1; i <= M ; i++){
		const [a, b] = input[i].split(' ').map(Number);
		if(union(a, b)) cnt--;
	}
	
	console.log(cnt);
	
});
